src/lava/lib/optimization/solvers/generic/scif/models.py

```diff
@@ -16,145 +17,176 @@
     QuboScif, Boltzmann
 
 
-@implements(proc=CspScif, protocol=LoihiProtocol)
-@requires(CPU)
-@tag('fixed_pt')
-class PyModelCspScifFixed(PyLoihiProcessModel):
+class PyModelAbstractScifFixed(PyLoihiProcessModel):
     """Fixed point implementation of Stochastic Constraint Integrate and
     Fire (SCIF) neuron for solving CSP problems.
     """
     a_in = LavaPyType(PyInPort.VEC_DENSE, int, precision=8)
     s_sig_out = LavaPyType(PyOutPort.VEC_DENSE, int, precision=8)
     s_wta_out = LavaPyType(PyOutPort.VEC_DENSE, int, precision=8)
 
+    cnstr_intg: np.ndarray = LavaPyType(np.ndarray, int, precision=24)
     state: np.ndarray = LavaPyType(np.ndarray, int, precision=24)
     spk_hist: np.ndarray = LavaPyType(np.ndarray, int, precision=8)
 
     step_size: np.ndarray = LavaPyType(np.ndarray, int, precision=24)
     theta: np.ndarray = LavaPyType(np.ndarray, int, precision=24)
-    noise_ampl: np.ndarray = LavaPyType(np.ndarray, int, precision=1)
-    cnstr_intg: np.ndarray = LavaPyType(np.ndarray, int, precision=24)
-    neg_tau_ref: np.ndarray = LavaPyType(np.ndarray, int, precision=24)
+    noise_ampl: np.ndarray = LavaPyType(np.ndarray, int, precision=5)
+    noise_prec: np.ndarray = LavaPyType(np.ndarray, int, precision=5)
+    sustained_on_tau: np.ndarray = LavaPyType(np.ndarray, int, precision=24)
 
     def __init__(self, proc_params):
-        super(PyModelCspScifFixed, self).__init__(proc_params)
+        """
+        Notes
+        -----
+        **Noise model**:
+        1. An unsigned 24-bit random number is chosen, say `R`
+        2. The bit-precision of `noise_amplitude` is inferred from its value.
+           Say, it is `ap`.
+        3. The number `R` is truncated to (`noise_precision` - `ap`) bits and
+           multiplied by `noise_amplitude`:
+           ```
+           S = noise_amplitude * [R << (24 - noise_precision + ap)] >> (
+           24 - noise_precision + ap)
+           ```
+        4. The number `S` is used in the SCIF dynamics.
+        """
+        super(PyModelAbstractScifFixed, self).__init__(proc_params)
         self.a_in_data = np.zeros(proc_params['shape'])
 
-    def _prng(self, intg_idx_):
-        """Pseudo-random number generator
+    @staticmethod
+    def _get_precision(vn_in: npty.NDArray) -> ty.Union[npty.NDArray, int]:
```

Comment:
Why don't you use the .bit_length() methods of the int class?

https://docs.python.org/3/library/stdtypes.html#int.bit_length